server/src/main/java/org/elasticsearch/action/bulk/BulkPrimaryExecutionContext.java

```diff
@@ -110,6 +112,10 @@ public int getUpdateRetryCounter() {
         return updateRetryCounter;
     }
 
+    public int getMappingUpdateRetryCounter() {
```

Comment:
We only use this number in one spot where we compare it with `0`, why not have a boolean flag in place here? Though, see my other comment. Can't we selectively retry when we were overtaken by some concurrent operation and need to now add more fields to `_ignored` instead of just retrying once without looking at the details of the mapping exception?

Reply:
I don't think that we can determine just from looking at the exception that it has been caused by a concurrent operation.

The exception is thrown here:
https://github.com/elastic/elasticsearch/blob/154dc9ac576035727b6f21586ce0a81091eaee99/server/src/main/java/org/elasticsearch/index/mapper/MappingLookup.java#L273-L282

This exception can also happen if we under-estimate the number of total fields that are added due to dynamic mappers in `DocumentParserContext#addDynamicMapper(Mapper mapper)`. In that case, instead of adding the fields to `_ignored` `addDynamicMapper`, we try a mapping update that does an authoritative validation on the number of field mappers. I don't see a way to differ between that scenario and the same exception being caused by a concurrent operation.

Reply:
I've changed this to a boolean now. I'm now also only retrying if the mapping version has changed.